From 61b76c859acfc383cd2bbde5448e93bc1beb16d0 Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Mon, 12 Mar 2018 18:49:16 -0400
Subject: [PATCH 196/196] [tmp] prueth: enable untagged vlan frames to
 workaround node table issue

Currently if vlan filtering is enabled, PRU stops sending supervision (SV)
frames to host since VLAN filtering filter out the untagged SV frames.
As usage of VLAN tag with SV frame is not yet supported, and the
requirement needs clarity, for now enable untagged VLAN frames to Host
by configuring the same in PRU. Without this, node table functionality
is broken. So add this workaround to resolve the issue for now.

Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
---
 drivers/net/ethernet/ti/prueth.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/ti/prueth.c b/drivers/net/ethernet/ti/prueth.c
index 96c4ed4..e802774 100644
--- a/drivers/net/ethernet/ti/prueth.c
+++ b/drivers/net/ethernet/ti/prueth.c
@@ -3238,12 +3238,18 @@ static int emac_add_del_vid(struct prueth_emac *emac,
 		writeb(val, sram + VLAN_FLTR_TBL_BASE_ADDR + index);
 	}
 
-	/* Enable VLAN filter for HSR/PRP. By default drop untagged frames
-	 * and allow priority tagged frames. Convention is 0 for allow,
-	 * 1 for drop.
+	/* Enable VLAN filter for HSR/PRP. By default, allow priority
+	 * tagged frames to be forwarded to host by enabling the
+	 * corresponding control bit. However this caused SV untagged
+	 * frames to be dropped as well which is undesirable. To
+	 * workaround that, enable untagged frames as well to host.
+	 * By convention, 0 in control bit means forward and 1 means
+	 * drop.
+	 *
+	 * TODO: Revisit this when tagged VLAN frames are supported
+	 * for Supervision frames.
 	 */
-	writeb(VLAN_FLTR_ENA | BIT(VLAN_FLTR_UNTAG_HOST_RCV_CTRL_SHIFT),
-	       sram + VLAN_FLTR_CTRL_BYTE);
+	writeb(VLAN_FLTR_ENA, sram + VLAN_FLTR_CTRL_BYTE);
 	spin_unlock_irqrestore(&emac->addr_lock, flags);
 
 	return 0;
-- 
2.7.4

